File: laurus/examples/lexical_search.rs
//! Lexical Search Example — all query types via Engine API.
//!
//! Demonstrates every lexical query type Laurus supports:
//!
//! 1. **TermQuery**         — exact single-term matching
//! 2. **PhraseQuery**       — exact word sequence matching
//! 3. **FuzzyQuery**        — approximate matching (typo tolerance)
//! 4. **WildcardQuery**     — pattern matching with `*` and `?`
//! 5. **NumericRangeQuery**  — numeric range filtering
//! 6. **GeoQuery**          — geographic radius / bounding box (Builder API only)
//! 7. **BooleanQuery**      — AND / OR / NOT combinations
//! 8. **SpanQuery**         — positional / proximity search (Builder API only)
//!
//! Each query type (where supported) includes both a Builder API example
//! and a QueryParser DSL example showing the equivalent text syntax.
//!
//! Run with: `cargo run --example lexical_search`

mod common;

use laurus::lexical::core::field::{BooleanOption, FloatOption, GeoOption, IntegerOption};
use laurus::lexical::span::{SpanQueryBuilder, SpanQueryWrapper};
use laurus::lexical::{
    BooleanQuery, FuzzyQuery, GeoQuery, LexicalQueryParser, NumericRangeQuery, PhraseQuery,
    TermQuery, TextOption, WildcardQuery,
};
use laurus::{
    DataValue, Document, Engine, LexicalSearchQuery, Result, Schema, SearchRequestBuilder,
};

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Laurus Lexical Search Example ===\n");

    // ─── Setup ───────────────────────────────────────────────────────────
    let storage = common::memory_storage()?;
    let analyzer = common::per_field_analyzer(&["category", "filename"]);

    let schema = Schema::builder()
        .add_text_field("title", TextOption::default())
        .add_text_field("body", TextOption::default())
        .add_text_field("category", TextOption::default())
        .add_text_field("filename", TextOption::default())
        .add_boolean_field("in_print", BooleanOption::default())
        .add_float_field("price", FloatOption::default())
        .add_integer_field("year", IntegerOption::default())
        .add_geo_field("location", GeoOption::default())
        .add_default_field("body")
        .build();

    let engine = Engine::builder(storage, schema)
        .analyzer(analyzer.clone())
        .build()
        .await?;

    // ─── Index documents ─────────────────────────────────────────────────
    let docs = vec![
        ("book1", Document::builder()
            .add_text("title", "The Rust Programming Language")
            .add_text("body", "Rust is a systems programming language focused on safety, speed, and concurrency")
            .add_text("category", "programming")
            .add_text("filename", "rust_book.pdf")
            .add_boolean("in_print", true)
            .add_field("price", DataValue::Float64(49.99))
            .add_integer("year", 2019)
            .add_field("location", DataValue::Geo(laurus::lexical::GeoPoint::new(37.7749, -122.4194))) // San Francisco
            .build()),
        ("book2", Document::builder()
            .add_text("title", "Python for Data Science")
            .add_text("body", "Python is a versatile programming language widely used in data science and machine learning")
            .add_text("category", "data-science")
            .add_text("filename", "python_data.epub")
            .add_boolean("in_print", true)
            .add_field("price", DataValue::Float64(39.99))
            .add_integer("year", 2021)
            .add_field("location", DataValue::Geo(laurus::lexical::GeoPoint::new(40.7128, -74.0060))) // New York
            .build()),
        ("book3", Document::builder()
            .add_text("title", "JavaScript Web Development")
            .add_text("body", "JavaScript powers the modern web from frontend frameworks to backend services")
            .add_text("category", "web-development")
            .add_text("filename", "javascript_web.pdf")
            .add_boolean("in_print", true)
            .add_field("price", DataValue::Float64(54.99))
            .add_integer("year", 2022)
            .add_field("location", DataValue::Geo(laurus::lexical::GeoPoint::new(51.5074, -0.1278))) // London
            .build()),
        ("book4", Document::builder()
            .add_text("title", "Machine Learning Algorithms")
            .add_text("body", "Understanding algorithms used in machine learning and artificial intelligence applications")
            .add_text("category", "data-science")
            .add_text("filename", "ml_algorithms.docx")
            .add_boolean("in_print", true)
            .add_field("price", DataValue::Float64(72.99))
            .add_integer("year", 2020)
            .add_field("location", DataValue::Geo(laurus::lexical::GeoPoint::new(37.4419, -122.1430))) // Palo Alto
            .build()),
        ("book5", Document::builder()
            .add_text("title", "Database Design Principles")
            .add_text("body", "Learn database design, SQL queries, and data management for modern applications")
            .add_text("category", "database")
            .add_text("filename", "db_design.pdf")
            .add_boolean("in_print", false)
            .add_field("price", DataValue::Float64(45.50))
            .add_integer("year", 2018)
            .add_field("location", DataValue::Geo(laurus::lexical::GeoPoint::new(47.6062, -122.3321))) // Seattle
            .build()),
        ("book6", Document::builder()
            .add_text("title", "The quick brown fox")
            .add_text("body", "The quick brown fox jumped over the lazy dog in a sunny meadow")
            .add_text("category", "fiction")
            .add_text("filename", "fox_story.txt")
            .add_boolean("in_print", false)
            .add_field("price", DataValue::Float64(12.99))
            .add_integer("year", 2023)
            .add_field("location", DataValue::Geo(laurus::lexical::GeoPoint::new(34.0522, -118.2437))) // Los Angeles
            .build()),
    ];

    println!("  Indexing {} documents...", docs.len());
    for (id, doc) in &docs {
        engine.add_document(id, doc.clone()).await?;
    }
    engine.commit().await?;
    println!("  Done.\n");

    // ─── Query parser for DSL examples ──────────────────────────────────
    let parser = LexicalQueryParser::new(analyzer).with_default_field("body");

    // =====================================================================
    // PART 1: TermQuery — exact single-term matching
    // =====================================================================
    println!("{}", "=".repeat(60));
    println!("PART 1: TermQuery");
    println!("{}", "=".repeat(60));

    println!("\n[1a] Search for 'rust' in body:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(TermQuery::new(
                    "body", "rust",
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[1b] Search for 'programming' in category (KeywordAnalyzer — exact):");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(TermQuery::new(
                    "category",
                    "programming",
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[1c] Search for in_print=true (boolean field):");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(TermQuery::new(
                    "in_print", "true",
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[1d] DSL: 'body:rust'");
    let query = parser.parse("body:rust")?;
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(query))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 2: PhraseQuery — exact word sequence
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 2: PhraseQuery");
    println!("{}", "=".repeat(60));

    println!("\n[2a] Phrase 'machine learning' in body:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(PhraseQuery::new(
                    "body",
                    vec!["machine".into(), "learning".into()],
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[2b] Phrase 'systems programming language' in body:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(PhraseQuery::new(
                    "body",
                    vec!["systems".into(), "programming".into(), "language".into()],
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[2c] DSL: 'body:\"machine learning\"'");
    let query = parser.parse("body:\"machine learning\"")?;
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(query))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 3: FuzzyQuery — approximate matching (typo tolerance)
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 3: FuzzyQuery");
    println!("{}", "=".repeat(60));

    println!("\n[3a] Fuzzy 'programing' (missing 'm', edit distance 2):");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(
                    FuzzyQuery::new("body", "programing").max_edits(2),
                )))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[3b] Fuzzy 'javascritp' (transposed, edit distance 1):");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(
                    FuzzyQuery::new("body", "javascritp").max_edits(1),
                )))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[3c] DSL: 'programing~2'");
    let query = parser.parse("programing~2")?;
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(query))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 4: WildcardQuery — pattern matching with * and ?
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 4: WildcardQuery");
    println!("{}", "=".repeat(60));

    println!("\n[4a] Wildcard '*.pdf' in filename:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(WildcardQuery::new(
                    "filename", "*.pdf",
                )?)))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[4b] Wildcard 'pro*' in body:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(WildcardQuery::new(
                    "body", "pro*",
                )?)))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[4c] DSL: 'body:pro*'");
    let query = parser.parse("body:pro*")?;
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(query))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 5: NumericRangeQuery — numeric range filtering
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 5: NumericRangeQuery");
    println!("{}", "=".repeat(60));

    println!("\n[5a] Books with price $40–$60:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(
                    NumericRangeQuery::f64_range("price", Some(40.0), Some(60.0)),
                )))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[5b] Books published after 2020:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(
                    NumericRangeQuery::i64_range("year", Some(2021), None),
                )))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[5c] DSL: 'price:[40 TO 60]'");
    let query = parser.parse("price:[40 TO 60]")?;
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(query))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 6: GeoQuery — geographic search (Builder API only)
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 6: GeoQuery (Builder API only — no DSL equivalent)");
    println!("{}", "=".repeat(60));

    println!("\n[6a] Within 100km of San Francisco (37.77, -122.42):");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(GeoQuery::within_radius(
                    "location", 37.7749, -122.4194, 100.0,
                )?)))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[6b] Bounding box — US West Coast (33, -123) to (48, -117):");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(
                    GeoQuery::within_bounding_box("location", 33.0, -123.0, 48.0, -117.0)?,
                )))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 7: BooleanQuery — AND / OR / NOT combinations
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 7: BooleanQuery");
    println!("{}", "=".repeat(60));

    println!("\n[7a] AND: 'programming' in body AND category='data-science':");
    let mut bq = BooleanQuery::new();
    bq.add_must(Box::new(TermQuery::new("body", "programming")));
    bq.add_must(Box::new(TermQuery::new("category", "data-science")));
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(bq)))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[7b] OR: category='programming' OR category='web-development':");
    let mut bq = BooleanQuery::new();
    bq.add_should(Box::new(TermQuery::new("category", "programming")));
    bq.add_should(Box::new(TermQuery::new("category", "web-development")));
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(bq)))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[7c] NOT: 'programming' in body, NOT 'python':");
    let mut bq = BooleanQuery::new();
    bq.add_must(Box::new(TermQuery::new("body", "programming")));
    bq.add_must_not(Box::new(TermQuery::new("body", "python")));
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(bq)))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[7d] DSL: '+body:programming -body:python'");
    let query = parser.parse("+body:programming -body:python")?;
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(query))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // =====================================================================
    // PART 8: SpanQuery — positional / proximity search (Builder API only)
    // =====================================================================
    println!("\n{}", "=".repeat(60));
    println!("PART 8: SpanQuery (Builder API only — no DSL equivalent)");
    println!("{}", "=".repeat(60));

    let sb = SpanQueryBuilder::new("body");

    println!("\n[8a] SpanNear: 'quick' near 'fox' (slop=1, ordered):");
    let span_near = sb.near(
        vec![Box::new(sb.term("quick")), Box::new(sb.term("fox"))],
        1,
        true,
    );
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(SpanQueryWrapper::new(
                    Box::new(span_near),
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\n[8b] SpanContaining: 'quick..fox' containing 'brown':");
    let big = sb.near(
        vec![Box::new(sb.term("quick")), Box::new(sb.term("fox"))],
        1,
        true,
    );
    let little = sb.term("brown");
    let containing = sb.containing(Box::new(big), Box::new(little));
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .lexical_query(LexicalSearchQuery::Obj(Box::new(SpanQueryWrapper::new(
                    Box::new(containing),
                ))))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    println!("\nLexical search example completed successfully!");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lexical_search_example() {
        let result = main();
        assert!(result.is_ok(), "lexical_search failed: {:?}", result.err());
    }
}

File: laurus-php/examples/lexical_search.php
<?php

// Lexical Search Example — all query types.
//
// Demonstrates every lexical query type Laurus supports:
//
// 1. TermQuery         — exact single-term matching
// 2. PhraseQuery       — exact word sequence matching
// 3. FuzzyQuery        — approximate matching (typo tolerance)
// 4. WildcardQuery     — pattern matching with * and ?
// 5. NumericRangeQuery — numeric range filtering (int and float)
// 6. GeoDistanceQuery / GeoBoundingBoxQuery — geographic radius / bounding box
// 7. BooleanQuery      — AND / OR / NOT combinations
// 8. SpanQuery         — positional / proximity search
//
// Usage:
//   cd laurus-php
//   cargo build --release
//   php -d extension=target/release/liblaurus_php.so examples/lexical_search.php

use Laurus\BooleanQuery;
use Laurus\FuzzyQuery;
use Laurus\GeoBoundingBoxQuery;
use Laurus\GeoDistanceQuery;
use Laurus\Index;
use Laurus\NumericRangeQuery;
use Laurus\PhraseQuery;
use Laurus\Schema;
use Laurus\SpanQuery;
use Laurus\TermQuery;
use Laurus\WildcardQuery;

function print_results(array $results): void
{
    if (empty($results)) {
        echo "  (no results)" . PHP_EOL;
        return;
    }
    foreach ($results as $r) {
        $doc = $r->getDocument();
        $title = $doc["title"] ?? "";
        printf("  id=%-10s  score=%.4f  title=%s\n", "'{$r->getId()}'", $r->getScore(), "'{$title}'");
    }
}

echo "=== Laurus Lexical Search Example ===" . PHP_EOL . PHP_EOL;

// ── Setup ──────────────────────────────────────────────────────────────
$schema = new Schema();
$schema->addTextField("title");
$schema->addTextField("body");
$schema->addTextField("category", true, true, false, "keyword");
$schema->addTextField("filename", true, true, false, "keyword");
$schema->addBooleanField("maintained");
$schema->addFloatField("rating");
$schema->addIntegerField("year");
$schema->addGeoField("location");
$schema->setDefaultFields(["body"]);

$index = new Index(null, $schema);

// ── Index documents ────────────────────────────────────────────────────
$docs = [
    ["laravel", [
        "title" => "Laravel Framework",
        "body" => "Laravel is an elegant PHP web application framework with expressive syntax and rich ecosystem",
        "category" => "framework",
        "filename" => "laravel_guide.pdf",
        "maintained" => true,
        "rating" => 4.8,
        "year" => 2011,
        "location" => ["lat" => 37.7749, "lon" => -122.4194], // San Francisco
    ]],
    ["symfony", [
        "title" => "Symfony Components",
        "body" => "Symfony is a set of reusable PHP components and a web application framework used by Laravel and Drupal",
        "category" => "framework",
        "filename" => "symfony_docs.epub",
        "maintained" => true,
        "rating" => 4.6,
        "year" => 2005,
        "location" => ["lat" => 48.8566, "lon" => 2.3522], // Paris
    ]],
    ["wordpress", [
        "title" => "WordPress Development",
        "body" => "WordPress powers over 40 percent of the web with its flexible plugin and theme architecture",
        "category" => "cms",
        "filename" => "wordpress_dev.pdf",
        "maintained" => true,
        "rating" => 4.2,
        "year" => 2003,
        "location" => ["lat" => 37.7749, "lon" => -122.4194], // San Francisco
    ]],
    ["phpunit", [
        "title" => "PHPUnit Testing Guide",
        "body" => "PHPUnit is the standard testing framework for PHP applications with assertions and mocking support",
        "category" => "testing",
        "filename" => "phpunit_manual.docx",
        "maintained" => true,
        "rating" => 4.5,
        "year" => 2004,
        "location" => ["lat" => 52.5200, "lon" => 13.4050], // Berlin
    ]],
    ["composer", [
        "title" => "Composer Package Manager",
        "body" => "Composer is a dependency management tool for PHP that manages libraries and autoloading",
        "category" => "tooling",
        "filename" => "composer_docs.pdf",
        "maintained" => true,
        "rating" => 4.7,
        "year" => 2012,
        "location" => ["lat" => 52.3676, "lon" => 4.9041], // Amsterdam
    ]],
    ["pear", [
        "title" => "PEAR Package Repository",
        "body" => "PEAR was the original PHP extension and application repository before Composer replaced it",
        "category" => "tooling",
        "filename" => "pear_archive.txt",
        "maintained" => false,
        "rating" => 2.5,
        "year" => 1999,
        "location" => ["lat" => 51.5074, "lon" => -0.1278], // London
    ]],
];

echo "  Indexing " . count($docs) . " documents..." . PHP_EOL;
foreach ($docs as [$docId, $doc]) {
    $index->addDocument($docId, $doc);
}
$index->commit();
echo "  Done." . PHP_EOL . PHP_EOL;

// =====================================================================
// PART 1: TermQuery — exact single-term matching
// =====================================================================
echo str_repeat("=", 60) . PHP_EOL;
echo "PART 1: TermQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[1a] Search for 'php' in body:" . PHP_EOL;
print_results($index->search(new TermQuery("body", "php"), 5));

echo PHP_EOL . "[1b] Search for 'framework' in category (exact):" . PHP_EOL;
print_results($index->search(new TermQuery("category", "framework"), 5));

echo PHP_EOL . "[1c] Search for maintained=true (boolean field):" . PHP_EOL;
print_results($index->search(new TermQuery("maintained", "true"), 5));

echo PHP_EOL . "[1d] DSL: 'body:php':" . PHP_EOL;
print_results($index->search("body:php", 5));

// =====================================================================
// PART 2: PhraseQuery — exact word sequence
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 2: PhraseQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[2a] Phrase 'web application framework' in body:" . PHP_EOL;
print_results($index->search(new PhraseQuery("body", ["web", "application", "framework"]), 5));

echo PHP_EOL . "[2b] Phrase 'dependency management' in body:" . PHP_EOL;
print_results($index->search(new PhraseQuery("body", ["dependency", "management"]), 5));

echo PHP_EOL . "[2c] DSL: 'body:\"web application framework\"':" . PHP_EOL;
print_results($index->search('body:"web application framework"', 5));

// =====================================================================
// PART 3: FuzzyQuery — approximate matching (typo tolerance)
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 3: FuzzyQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[3a] Fuzzy 'laravl' (missing 'e', max_edits=2):" . PHP_EOL;
print_results($index->search(new FuzzyQuery("body", "laravl", 2), 5));

echo PHP_EOL . "[3b] Fuzzy 'sympfony' (extra 'p', max_edits=1):" . PHP_EOL;
print_results($index->search(new FuzzyQuery("body", "sympfony", 1), 5));

echo PHP_EOL . "[3c] DSL: 'laravl~2':" . PHP_EOL;
print_results($index->search("laravl~2", 5));

// =====================================================================
// PART 4: WildcardQuery — pattern matching with * and ?
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 4: WildcardQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[4a] Wildcard '*.pdf' in filename:" . PHP_EOL;
print_results($index->search(new WildcardQuery("filename", "*.pdf"), 5));

echo PHP_EOL . "[4b] Wildcard 'php*' in body:" . PHP_EOL;
print_results($index->search(new WildcardQuery("body", "php*"), 5));

echo PHP_EOL . "[4c] DSL: 'body:php*':" . PHP_EOL;
print_results($index->search("body:php*", 5));

// =====================================================================
// PART 5: NumericRangeQuery — numeric range filtering
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 5: NumericRangeQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[5a] Rating 4.5-5.0 (float range):" . PHP_EOL;
print_results($index->search(new NumericRangeQuery("rating", 4.5, 5.0, "float"), 5));

echo PHP_EOL . "[5b] Projects started from 2010 onwards (integer range):" . PHP_EOL;
print_results($index->search(new NumericRangeQuery("year", 2010, null), 5));

echo PHP_EOL . "[5c] DSL: 'rating:[4.5 TO 5.0]':" . PHP_EOL;
print_results($index->search("rating:[4.5 TO 5.0]", 5));

// =====================================================================
// PART 6: GeoDistanceQuery / GeoBoundingBoxQuery — geographic search
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 6: GeoDistanceQuery / GeoBoundingBoxQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[6a] Within 100 km of Paris (48.86, 2.35):" . PHP_EOL;
print_results($index->search(GeoDistanceQuery::withinRadius("location", 48.8566, 2.3522, 100_000.0), 5));

echo PHP_EOL . "[6b] Bounding box — Europe (47, -1) to (53, 14):" . PHP_EOL;
print_results($index->search(GeoBoundingBoxQuery::withinBoundingBox("location", 47.0, -1.0, 53.0, 14.0), 5));

// =====================================================================
// PART 7: BooleanQuery — AND / OR / NOT combinations
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 7: BooleanQuery" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[7a] AND: 'php' in body AND category='framework':" . PHP_EOL;
$bq = new BooleanQuery();
$bq->must(new TermQuery("body", "php"));
$bq->must(new TermQuery("category", "framework"));
print_results($index->search($bq, 5));

echo PHP_EOL . "[7b] OR: category='framework' OR category='cms':" . PHP_EOL;
$bq = new BooleanQuery();
$bq->should(new TermQuery("category", "framework"));
$bq->should(new TermQuery("category", "cms"));
print_results($index->search($bq, 5));

echo PHP_EOL . "[7c] NOT: 'php' in body, NOT 'composer':" . PHP_EOL;
$bq = new BooleanQuery();
$bq->must(new TermQuery("body", "php"));
$bq->mustNot(new TermQuery("body", "composer"));
print_results($index->search($bq, 5));

echo PHP_EOL . "[7d] DSL: '+body:php -body:composer':" . PHP_EOL;
print_results($index->search("+body:php -body:composer", 5));

// =====================================================================
// PART 8: SpanQuery — positional / proximity search
// =====================================================================
echo PHP_EOL . str_repeat("=", 60) . PHP_EOL;
echo "PART 8: SpanQuery (no DSL equivalent)" . PHP_EOL;
echo str_repeat("=", 60) . PHP_EOL;

echo PHP_EOL . "[8a] SpanNear: 'plugin' near 'theme' (slop=2, ordered):" . PHP_EOL;
$spanQ = SpanQuery::near("body", ["plugin", "theme"], 2, true);
print_results($index->search($spanQ, 5));

echo PHP_EOL . "[8b] SpanContaining: 'web..framework' containing 'application':" . PHP_EOL;
$big = SpanQuery::near("body", ["web", "framework"], 2, true);
$little = SpanQuery::term("body", "application");
$containing = SpanQuery::containing("body", $big, $little);
print_results($index->search($containing, 5));

echo PHP_EOL . "Lexical search example completed!" . PHP_EOL;
